Feature/caitlin kunchur

ckunchur opened the pull request:
<!--

This source file is part of the Stanford CS342 - Building for Digital Health class

SPDX-FileCopyrightText: 2022 Stanford University

SPDX-License-Identifier: MIT

-->

# Submitting class profile view 

## :recycle: Current situation & Problem
The task was to create a personalized view containing information about me. No issues except slightly slow rendering due to the number of photos I embedded; not sure how to improve this, as I already compressed the images as much as possible.

## :bulb: Proposed solution
Successfully built new view and linked it to the main page of the application.

## :gear: Release Notes 
My view scrolls vertically and contains 3 portions. The top portion has my picture and basic information about my name, school, and class. It also contains a row of social media icon buttons that link to my social media pages. I overlayed this data on a background image. The middle portion is a short description of my research interests and hobbies. The bottom half of the view contains three horizontally scrolling views containing information on the recent books I read, the albums I listened to, and my top hobbies. There is also an inspirational quote and an image positioned on the bottom of the outer view. 


## :heavy_plus_sign: Additional Information
*Provide some additional information if possible*

I drew inspiration from the SwiftUI Landmarks Tutorial for creating the horizontally scrolling views. All other component details were customized based on reading Swift documentation. 

### Related PRs
*Reference the related PRs*

### Testing
*Are there tests included? If yes, which situations are tested, and which corner cases are missing?*

The tests currently check that the app launches and that my profile button from the main screen works. It would be wise to add additional tests to check that the social media buttons work. 

### Reviewer Nudging
*Where should the reviewer start? Where is a good entry point?*

### Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/CS342/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/CS342/.github/blob/main/CONTRIBUTING.md):
- [+] I agree to follow the [Code of Conduct](https://github.com/CS342/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/CS342/.github/blob/main/CONTRIBUTING.md).

ckunchur committed:
made final edits to profile view

ckunchur committed:
edited UI test file

ckunchur committed:
added REUSE header to each of my files

ckunchur committed:
added license files for all images used in view

ckunchur committed:
Merge branch 'main' into feature/CaitlinKunchur

PSchmiedmayer committed:
Merge branch 'main' into feature/CaitlinKunchur

ckunchur committed:
Fixed style and edited package file

ckunchur committed:
added more files

PSchmiedmayer commented on CS342MemberViews/Sources/CaitlinKunchur/CaitlinKunchur.swift:
```diff
@@ -0,0 +1,57 @@
+                    ReadsRow().offset(y: -150)
+                    ListensRow().offset(y: -150)
+                    HobbiesRow().offset(y: -150)
+                    Text("\"Dream big and dare to fail.\"").offset(y: -80).font(.custom(
+                        "Georgia-Italic",
+                        fixedSize: 25)).foregroundColor(.white)
```
As noted in the SwiftLint errors, please try to put each modifier in a new line.

I would suggest a formatting like this:
```suggestion
                    Text("\"Dream big and dare to fail.\"")
                        .offset(y: -80)
                        .font(
                            .custom(
                                "Georgia-Italic",
                                fixedSize: 25
                            )
                        )
                        .foregroundColor(.white)
```

PSchmiedmayer commented on CS342MemberViews/Sources/CaitlinKunchur/HobbiesRow.swift:
```diff
@@ -0,0 +1,89 @@
+import SwiftUI
+
+
+public struct HobbiesRow: View {
```
Very nice job displaying all your hobbies in your view!! 👍

Please be sure that you address all the SwiftLint errors in this file before you merge the PR.

PSchmiedmayer commented on CS342MemberViews/Sources/CaitlinKunchur/HobbiesRow.swift:
```diff
@@ -0,0 +1,89 @@
+                                .foregroundColor(.primary)
+                                .font(.caption).padding([.bottom], 10)
+                        }
+                        .padding(.leading, 18)
```
I would suggest you to move the different subviews into a separate private computed property to make the overall view body smaller and easier to read. E.g. you can move this view to a private property like this:
```swift
private var photographyView: some View {
    VStack {
         Image(systemName: "camera.circle")
             .resizable()
             .frame(width: 35, height: 35)
             .cornerRadius(5)
         Text("Photography")
             .foregroundColor(.primary)
             .font(.caption).padding([.bottom], 10)
    }
         .padding(.leading, 18)
}
```

One even better solution would be using a `ForEach` loop and a model to iterate over your views to generate them:

1. Generate a model type:
```swift
struct Hobby: Identifiable {
    let imageSystemName: String
    let text: String

    var id: String {
        text
    }
}
```

2. Create a collection of hobbies in your view:
```swift
    let myHobbies = [
        Hobby(imageSystemName: "camera.circle", text: "Photography"),
         Hobby(imageSystemName: "...", text: "..."),
    ]
```

3. Now use a ForEach to create your views:
```swift
        ForEach(myHobbies) { myHobby in 
             VStack {
                   Image(systemName: myHobby.imageSystemName)
                       .resizable()
                       .frame(width: 35, height: 35)
                       .cornerRadius(5)
                   Text(myHobby.text)
                       .foregroundColor(.primary)
                       .font(.caption).padding([.bottom], 10)
               }
                  .padding(.leading, 18)
        }
```

PSchmiedmayer commented on CS342MemberViews/Sources/CaitlinKunchur/ListensRow.swift:
```diff
@@ -0,0 +1,93 @@
+                ScrollView(.horizontal, showsIndicators: false) {
+                    HStack {
+                        VStack {
+                            Bundle.module.image(fromFileNamed: "car-cover")
```
Similar to the `HobbiesRow.swift`, I would suggest moving the views into smaller chunks or using a `ForEach` loop to create your views.

ckunchur committed:
fixed conflicts

PSchmiedmayer commented on CS342MemberViews/Sources/CaitlinKunchur/ProfileRow.swift:
```diff
@@ -0,0 +1,46 @@
+                Spacer()
+            } .offset(y: -150)
+            
+            Text("Caitlin Kunchur is a junior at Stanford studying Biomedical Computation. She is a graphics and tech team staffer for the Stanford Daily and co-chair of the Stanford Students in Biodesign Connect committee. Her past research experience inlcudes automating large-scale structural phylogeny for drug repurposing and developing software for automated retinal image processing. Her curr ...
```
Please move longer sections of text into, e.g., a markdown file to focus on your code in the swift files and the content of your application in the resources folders.
You can take a look at e.g. my bio or @aalami5's bio in his view in the CS342 class application.

PSchmiedmayer commented on CS342MemberViews/Sources/CaitlinKunchur/ReadsRow.swift:
```diff
@@ -0,0 +1,93 @@
+                
+                ScrollView(.horizontal, showsIndicators: false) {
+                    HStack {
+                        VStack {
```
Similar to the `HobbiesRow.swift`, I would suggest moving the views into smaller chunks or using a `ForEach` loop to create your views.

ckunchur replied:
Hi Paul! Thanks for the feedback! Vishnu and I were able to implement basically all the changes you suggested above! I fixed as many of the SwiftLint errors as I could and just have these two showing up on my end:
<img width="542" alt="image" src="https://user-images.githubusercontent.com/67437371/215292314-d7222922-2d19-4f66-b6d4-0ed4b641761d.png">

However, after committing and pushing, I still see more than just these two swift lint errors on Github's end. I am also not sure how to resolve the "exclude" error because I do have a file called "CaitlinKunchur-bio.md.license"

PSchmiedmayer reviewed (changes requested):
Hi @ckunchur,
First of all: Great job with the PR description!! 🚀

One of the most creative solutions I have seen, you will definitely get some bonus points for that 🥳

I have added some comments on how to improve your code in the PR review.
The next step would be 
1. to fix the remaining SwiftLint errors locally by installing SwiftLint or going through the error messages in the PR diff here on GitHub. You can find out more on how to install SwiftLint on your machine here: https://github.com/realm/SwiftLint
2. After that please also fix the REUSE errors by adding a license to the `CS342MemberViews/Sources/CaitlinKunchur/Resources/CaitlinKunchur-bio.md` file or adding a `CaitlinKunchur-bio.md.license` file in the same folder the the appropriate copyright statement.

I have also merged the latest main state in your branch to make merging the PR easy as you have already resolved one merge conflict already.
Be sure that you `git pull` the latest state to your local version if you accept any of the suggestions here on GitHub and get the merge conflict to your local branch 🚀

ckunchur committed:
fixed more swiftlint errors

ckunchur committed:
Fixed more swiftlint errors

ckunchur committed:
renamed license file

ckunchur committed:
fixed license name in Package

ckunchur commented:
> Hi @ckunchur, First of all: Great job with the PR description!! 🚀
> 
> One of the most creative solutions I have seen, you will definitely get some bonus points for that 🥳
> 
> I have added some comments on how to improve your code in the PR review. The next step would be
> 
> 1. to fix the remaining SwiftLint errors locally by installing SwiftLint or going through the error messages in the PR diff here on GitHub. You can find out more on how to install SwiftLint on your machine here: https://github.com/realm/SwiftLint
> 2. After that please also fix the REUSE errors by adding a license to the `CS342MemberViews/Sources/CaitlinKunchur/Resources/CaitlinKunchur-bio.md` file or adding a `CaitlinKunchur-bio.md.license` file in the same folder the the appropriate copyright statement.
> 
> I have also merged the latest main state in your branch to make merging the PR easy as you have already resolved one merge conflict already. Be sure that you `git pull` the latest state to your local version if you accept any of the suggestions here on GitHub and get the merge conflict to your local branch 🚀

Hi Paul thank you for all the feedback!!! With Vishnu's help I have implemented most of the changes you suggested (foreach, abstracting the views a bit more, etc.)! REUSE tests pass, but I am still having a couple of SwiftLint errors that I cannot figure out how to fix. 
<img width="529" alt="image" src="https://user-images.githubusercontent.com/67437371/215293408-3870117d-e454-402c-a0be-20bceb8ad130.png">. I have a file called "CaitlinKunchur-bio.md.license" and it is correctly named as such in the Package file. As for the first ACL error, not sure how to interpret that either? 

Thanks!

vishnuravi committed:
Merge from

vishnuravi committed:
Refactor HobbiesRow.swift to fix SwiftLint error

PSchmiedmayer committed:
Merge branch 'main' into feature/CaitlinKunchur

PSchmiedmayer committed:
Update Tests
